MTM-64360 User guide for MQTT Service Pulsar client

scmi-c8y opened the pull request:
Opening this up for the full review now that we have a complete first draft plus work-in-progress sample code in https://github.com/Cumulocity-IoT/cumulocity-examples/pull/158.

The relevant part of the preview is: https://cumulocity-iot-c8y-docs-build-pr-3642.surge.sh/device-integration/mqtt-service/#pulsar-client

Some open questions that we should answer collectively, and make sure the implementation agrees with the doc:
* ~~Should a **missing** `client` property should be treated the same as empty, or as a bad message to be ignored?~~
* ~~I documented using the "routing key" (which is used to calculate the topic partition _and_ the key-shared partition if the ordering key is not set) rather than the "ordering key" (which is only used by key-shared subscriptions) because 1) then we only have to set one key, 2) the routing key can be set from a String (at least in the Java API) and 3) IMHO this is less confusing for the user. However if people disagree with that we can change the doc/example to use the ordering key.~~
* ~~Should we have a stronger statement about not supporting clients that don't set the key correctly?~~

To do:
- [x] Update code snippets based on the outcome of https://github.com/Cumulocity-IoT/cumulocity-examples/pull/158
- [x] Double-check URLs to the example code
- [x] Update version and date in the change log!
- [x] LLM review for consistency and conciseness?

scmi-c8y committed:
MTM-64360 Initial outline of MQTT Pulsar doc

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+
+* Connecting to Pulsar requires a URL and valid credentials
+* For microservices, use the URL given to the microservice in the `C8Y_PULSAR_URL` environment variable
+* For external applications, use `pulsar://<domain>:6651` (this endpoint uses TLS, client certificates not currently supported)
```
```suggestion
* For external applications, use `pulsar+ssl://<domain>:6651` (this endpoint uses TLS, client certificates not currently supported)
```

scmi-c8y replied:
Oops! Yes will certainly fix this.

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+<p>
+
+* Connections will be authorized to access only the Pulsar topics used by the MQTT Service, further controlled by roles
+* To consume messages from MQTT devices, the authenticated user must have `READ` permission on the `MQTT_SERVICE_MESSAGING_TOPICS` role
```
**_throught:_** For MS manifests, it may be useful to show the full enum name (e.g., `MQTT_SERVICE_MESSAGING_TOPICS_READ`) since they will need that.

scmi-c8y replied:
I was planning to show a snippet from a sample MS manifest that would have the full name - would that be OK?

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+Microservices and external applications can connect to the {{< product-c8y-iot >}} Messaging Service to receive messages from devices connected to the MQTT Service, and send messages to those devices.
+The Messaging Service is a modified deployment of Apache Pulsar, and your applications will use the Pulsar client protocol directly to connect to it.
+
+### Connecting to the Messaging Service
```
**_question:_** Would adding a short introduction to the usage of the Pulsar client, consumer, and producer to help familiarise the reader—preferably with a code snippet — be a good idea? After that, we can follow up with usage examples for the MQTT Service, like you have done.

scmi-c8y replied:
I don't know, I feel like that would be repeating material that is already in the Pulsar documentation. I did plan to have code snippets showing how to create a client, consumer & producer (and publish a message to the producer). If you stick all those together you have a trivial client example.

scmi-c8y replied:
I added a link to the Pulsar docs to the introduction - making it clear that you should read this to understand the basics of pub/sub messaging and how to use Pulsar. This doc only covers what you need to know to build a simple client. https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/68a87cbc33d7dde3fce15d6c47dbc354b2ef478b

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+Microservices and external applications can connect to the {{< product-c8y-iot >}} Messaging Service to receive messages from devices connected to the MQTT Service, and send messages to those devices.
+The Messaging Service is a modified deployment of Apache Pulsar, and your applications will use the Pulsar client protocol directly to connect to it.
+
+### Connecting to the Messaging Service
```
**_nitpick:_** A quick prerequisites section helps perhaps?

scmi-c8y replied:
Yes, good thought. What would you have as pre-requisites?

rohitjoshi-c8y replied:
These points are already described later (tenant access, Pulsar client library, credentials/roles, URL), but grouping them up front helps readers know what they need before diving in.

scmi-c8y replied:
Hopefully this is what you had in mind? https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/39cf92da425c0976673cb8609ab568a02215e19c

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+weight: 30
+layout: redirect
+title: Connecting microservices and applications
+---
```
**_thought:_** Could a simple architecture diagram showing how these components fit together help? Maybe too much to ask for now, and perhaps we can come back to it later.

scmi-c8y replied:
I was thinking about updating the diagrams we already have in the intro section of the doc. I guess I might be inspired to copy parts of those into the later sections and add more detail.

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+<p>
+
+* Code snippet showing how to publish
+
```
**_issue/suggestion:_** It might be useful to add a section on error handling so readers know how to deal with common error scenarios.

scmi-c8y replied:
Yes, that's another good idea. Not sure what I would put in it exactly... probably I should try to remember all the ways I mess up in writing some example code.

scmi-c8y replied:
I added some thoughts on this subject: https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/69e7aed1bf0c916c352919f0fc1478b7ffb044b3.

Let me know what you think, with your greater knowledge of the Pulsar client library. For example I know we generally throw away a `Producer` after any failure to publish, but is this also the right advice for a `Consumer`, or does it generally re-connect automatically?

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+* To consume messages from MQTT devices, the authenticated user must have `READ` permission on the `MQTT_SERVICE_MESSAGING_TOPICS` role
+* To publish messages to MQTT devices, the authenticated user must have `UPDATE` permission on the `MQTT_SERVICE_MESSAGING_TOPICS` role
+* For microservices, these permissions should be configured in the microservice manifest, and they will be applied to the service user
+* For external applications, these permissions should be configured on the authenticated user through the Administration application
```
**_question:_** Should we also highlight other security features, such as tenant isolation?

scmi-c8y replied:
That feels like something we would have already mentioned in the introduction section.

scmi-c8y replied:
Added a line about device isolation to the introduction, and it is mentioned again in the publishing section. https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/9ad07568be5573d4f3cbead38f28b4a171ad06b6

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+<p>
+
+* Code snippet showing how to shut down cleanly
+
```
**_throught:_** Would it be helpful to include a short “common pitfalls” list (e.g., forgetting to acknowledge messages -> backlog quota filled)?

scmi-c8y replied:
That is really what this "best practices" section was supposed to be, and I think I already mentioned the consequences of not acknowledging a few lines up from here. It should be clearer when all these bullets are turned into actual prose.

rohitjoshi-c8y reviewed (commented):
Just a few early comments for now, as requested — I will take a closer look later.

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Rohit Joshi <89976252+rohitjoshi-c8y@users.noreply.github.com>

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,86 @@
+### Publishing messages to MQTT devices
+
+* All messages to MQTT devices are published on a single Pulsar topic, `persistent://<tenant>/mqtt/to-device`
+* The topic to publish to must be set using the `topic` message property
```
AND I love these two protocols. 
```suggestion
* The MQTT topic to publish to must be set using the `topic` message property
```

The topic message property, not to be confused with the Producer topic setting. 😄

rthrippleton replied:
```suggestion
* The topic to publish to must be set using the `channel` message property
```

scmi-c8y committed:
MTM-64360 WIP on intro and connecting sections

scmi-c8y committed:
MTM-64360 WIP payloads and properties section

scmi-c8y committed:
MTM-64360 WIP consuming messages section

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+
+### Connecting to the Messaging Service
+
+To connect your client to the Messaging Service, you will need to use a [Pulsar client library](https://pulsar.apache.org/docs/4.0.x/client-libraries/).
```
Is it definitely ok/recommended to use a v4 client against the v3 server deployment we use?

rthrippleton replied:
Oh, never mind, they have a strong documented guarantee

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+The payload is a sequence of zero or more bytes, identical to the payload of the MQTT `PUBLISH` message that the Pulsar message corresponds to.
+It is the client's responsiblity to understand the format of the payloads produced and accepted by the MQTT devices it communicates with.
+
+Message properties are name-value pairs, where both the name and the value are text strings.
```
```suggestion
Pulsar message properties are name-value pairs, where both the name and the value are text strings.
```
Now, I'm not suggesting using "(Pulsar,MQTT) message" everywhere, but the text needs to periodically reassert the context. I'll leave it to your judgement how often we do that.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+The client will not be able to consume from any other topic.
+
+The client identifier of the device that published the messages, and the MQTT topic it was published on, can be obtained from the message properties `client` and `channel` as described above.
+This means that your client **must** consume every message published by every device connected to the MQTT Service for the tenant, event those you are not interested in.
```
```suggestion
This means that your client **must** consume every message published by every device connected to the MQTT Service for the tenant, even those you are not interested in.
```

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+* All messages to MQTT devices are published on a single Pulsar topic, `persistent://<tenant>/mqtt/to-device`
+* The topic to publish to must be set using the `topic` message property
+* Because MQTT Service devices are isolated from each other, it is also necessary to set the `clientid` property to the id of the target device
+* However, if you want to publish the message to all devices that are subscribed to the named topic, leave the `clientid` property empty (or missing?)
```
```suggestion
* Because MQTT Service devices are isolated from each other, it is also necessary to set the `client` property to the id of the target device
* However, if you want to publish the message to all devices that are subscribed to the named topic, leave the `client` property empty (or missing?)
```

scmi-c8y replied:
I think I re-wrote this section completely, hopefully it's OK now.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+
+### Publishing messages to MQTT devices
+
+* All messages to MQTT devices are published on a single Pulsar topic, `persistent://<tenant>/mqtt/to-device`
```
"are published" reads like it's the system doing it for us. Suggest something more like "should be published" or "You will publish ... "

scmi-c8y replied:
Re-written much more verbosely.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+* All messages to MQTT devices are published on a single Pulsar topic, `persistent://<tenant>/mqtt/to-device`
+* The topic to publish to must be set using the `topic` message property
+* Because MQTT Service devices are isolated from each other, it is also necessary to set the `clientid` property to the id of the target device
+* However, if you want to publish the message to all devices that are subscribed to the named topic, leave the `clientid` property empty (or missing?)
```
```suggestion
* However, if you want to publish the message to all devices that are subscribed to the named topic, do not set the `client` property.
```
I'll update this in the relevant ticket.

scmi-c8y replied:
To be fair, I made that up without asking anyone. We need to agree in the team what the behaviour should be and align the documentation with the implementation. I'm going to leave it open as a TBD for the full review.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+
+### Best practices to ensure reliable message delivery {#reliable-delivery-best-practices}
+
+* Topics have a "backlog quota" that affects the number of unacknowledged messages that can be outstanding
```
"Topics" ... you know what I'm going to say :-)

scmi-c8y replied:
Yeah, I think I made this clear enough in the text (remember you were just reviewing an outline) but you can nitpick it in the full review.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,252 @@
+
+* Code snippet showing how to publish
+
+### Best practices to ensure reliable message delivery {#reliable-delivery-best-practices}
```
Direction is ambiguous, which is significant as we don't currently deal with the other direction.
```suggestion
### Best practices to ensure reliable message delivery from devices {#reliable-delivery-best-practices}
```

scmi-c8y committed:
MTM-64360 WIP publishing messages section

scmi-c8y committed:
MTM-64360 Typo

scmi-c8y committed:
MTM-64360 Align snippets with the working example

scmi-c8y committed:
MTM-64360 WIP quotas, limits, best practices

scmi-c8y committed:
MTM-64360 Use the final agreed property names

scmi-c8y committed:
MTM-64360 Minor cleanup

scmi-c8y committed:
MTM-64360 Mention the example client

scmi-c8y committed:
MTM-64360 Mention device isolation in the intro

scmi-c8y committed:
MTM-64360 Use the Pulsar docs for pub/sub intro

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Richard Thrippleton <richard.thrippleton@softwareag.com>

scmi-c8y committed:
MTM-64360 Add a lovely diagram

scmi-c8y committed:
Merge branch 'develop' into feature/MTM-64360/mqtt-service-pulsar-client

scmi-c8y committed:
MTM-64360 Alt text for image

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,388 @@
+To do this, your microservice or external application will connect to the {{< product-c8y-iot >}} Messaging Service, a modified deployment of [Apache Pulsar](https://pulsar.apache.org/), and use the Pulsar protocol to publish and consume MQTT messages.
+The diagram below shows the important interfaces and data flows used when interacting with the MQTT Service through Pulsar.
+
+<p align="center" width="100%">
```
**suggestion:** I think this wonderful diagram should be moved/condensed/changed+reused for Overview / Architecture - the diagram there is very legacy-oriented. Maybe now, or maybe in a later ticket, not sure.

scmi-c8y replied:
I broadly agree, although I'm wondering if some of the details could be simplified for a top-level architecture diagram, and if IDP and/or core should be explicitly mentioned on the RHS of the diagram? What do you think?

rthrippleton replied:
Too early to mention IDP I think! I suppose we don't want to over-egg "Microservice or external client" in the Overview, but maybe "Cumulocity platform and customer application" would be a good substitute.

Other simplifications are to remove the smaller blocks in the middle of the flow like "Device 1 Topic A" and "Pulsar from-device topic". Then we're still giving a good overview, without being overly zoomed-in.

scmi-c8y replied:
Thanks. Will keep all this in mind when I re-work the Overview.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,388 @@
+The diagram below shows the important interfaces and data flows used when interacting with the MQTT Service through Pulsar.
+
+<p align="center" width="100%">
+    <img width="80%" src="/images/mqtt-service/mqtt-service-pulsar-connections.svg" alt="MQTT Service Pulsar connections">
```
**praise:** Wonderful diagram, just the right amount of fiction to help a user model it in their mind.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
**thought:** Now that we're exposing a reliable messaging system more explicitly, used to communicate with another (potentially) reliable messaging service, I wonder if there's scope for confusion. As a user I might think that when I publish a message to Pulsar, when I get an acknowledgement from it it means that the MQTT device has acknowledged it. A similar delusion in the other direction perhaps.

Somewhere we might want to disillusion them of this, and be clear about it being decoupled. And another consequence of decoupling is a lack of errors - publishing to a non-existent MQTT topic or device via Pulsar will just complete and drop the message into a blackhole.

scmi-c8y replied:
Yeah, I already had to explain this exact point to some field folks so it should be mentioned in here and probably also in the FAQ.

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/a5377068aa2f51a2a37c21c623443b1a675fd761

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,388 @@
+The message will only be published to a client that is connected at the time the MQTT Service processes the published message.
+
+In order to enforce device-level isolation, the message will be published **only** to the specific MQTT client identified by the `clientID` message property, provided that client has an active subscription to the relevant MQTT topic.
+If the `clientID` property is empty, the message will be sent to **all** connected MQTT clients with active subscriptions to the MQTT topic.
```
Or at least that's what https://cumulocity.atlassian.net/browse/MTM-64283 says.

```suggestion
If the `clientID` property is missing, the message will be sent to **all** connected MQTT clients with active subscriptions to the MQTT topic.
```

scmi-c8y replied:
Hmm, so it does. In that case, what should the behaviour be if the property is present but with an empty value, which I believe is allowed by the Pulsar protocol? Should that be treated the same as a missing property, or as an error? (and in that case how do we respond - the only real option we have is to silently drop the message, unless we want to introduce a DLQ or similar). I don't have a strong opinion other than that the doc should match the implementation.

rthrippleton replied:
Just the same as any other incorrect clientID. Empty is just a zero-length string that somebody has to have explicitly set:
* Properties are `{...}` - goes to all clients
* Properties are `{"clientID":"realclient"}` goes to one client
* Properties are `{"clientID":"nonexistentclient"}` gets dropped on the floor
* Properties are `{"clientID":""}` gets dropped on the floor (why did you do that to yourself? Stop hitting yourself.)

scmi-c8y replied:
Makes sense. Clarified in https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/7b189f58560a4414d41ee4ff61c9b544c6bc2676.

scmi-c8y committed:
MTM-64361 Add a change log entry

scmi-c8y committed:
MTM-64360 Clarify handling of clientID property

scmi-c8y committed:
MTM-64360 More clarification of property handling
Document the cases from MTM-64678 where invalid messages will be
dropped and alarms raised, and valid messages that still could be
silently dropped.

scmi-c8y committed:
MTM-64360 More clarifications and warnings

schmiel-cumulocity commented on content/change-logs/platform-services/mqtt-service-0.9.x-pulsar-client-preview.md:
```diff
@@ -0,0 +1,40 @@
+Direct Messaging Service connectivity for MQTT Service clients is now available as part of the MQTT Service Public Preview.
+
+All microservice and external application clients currently using the MQTT Service Java client SDK should migrate to the Pulsar client protocol as soon as possible.
+Detailed [documentation](/device-integration/mqtt-service/#pulsar-client) of the requirements, conventions and best practices for connecting Pulsar clients to the Messaging Service is available, along with [example code for an external client](https://github.com/Cumulocity-IoT/cumulocity-examples/tree/develop/mqtt-service/simple-pulsae-client).
```
```suggestion
Detailed [documentation](/device-integration/mqtt-service/#pulsar-client) of the requirements, conventions and best practices for connecting Pulsar clients to the Messaging Service is available, along with [example code for an external client](https://github.com/Cumulocity-IoT/cumulocity-examples/tree/develop/mqtt-service/simple-pulsar-client).
```

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+        final MessageListener<String> listener = new MessageListener<String>() {
+            @Override
+            public void received(Consumer<String> consumer, Message<String> message) {
+                final String clientId = message.getProperty("client");
```
```suggestion
                final String clientId = message.getProperty("clientID");
```

scmi-c8y replied:
All the code snippets will be updated after https://github.com/Cumulocity-IoT/cumulocity-examples/pull/158 is approved, but I will apply these suggestions anyway to avoid confusing PS folks who are already using this doc.

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+            @Override
+            public void received(Consumer<String> consumer, Message<String> message) {
+                final String clientId = message.getProperty("client");
+                final String topic = message.getProperty("channel");
```
```suggestion
                final String topic = message.getProperty("topic");
```

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+To consume messages from the topic, your client should create a Pulsar `Consumer` and subscribe it to the topic.
+The consumer should register a `MessageListener` callback that will be called whenever a new message arrives on the topic.
+The `MessageListener` implementation shows how to access the payload and properties of the received messages.
+For simplicity and clarity, the example assumes that message payloads are simple text strings.
```
MQTT Service always uses `<byte[]>` producer, so I would stick with it on the consumer side too (it may not even work as the producer schema is always `<byte[]>` and the consumer tries to use `<String>`).

rohitjoshi-c8y replied:
Yes, `<String>` does not work in our case.

scmi-c8y replied:
It actually did work, but probably only by good luck. Anyway I agree that we should use `byte[]` and `Schema.BYTES` to avoid any confusion especially if users copy/paste the example as they probably will. The code snippets are still wrong - they will be fixed when the example is approved and copy-pasted in here.

https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/382f350e65557ce4351ce3bd0ee8ff0f68962b6a

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+The message will only be published to a client with an active subscription to the named MQTT topic.
+The message will only be published to a client that is connected at the time the MQTT Service processes the published message.
+
+Successfully publishing a message to the Messaging Service does **not** mean that the message has been successfully delivered to any MQTT device.
```
Should we say something about QoS to avoid future questions?
AFAIR from Anthony's PR, currently we are pushing the message with the same QoS level as it was in the original subscription message from the device.

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/382f350e65557ce4351ce3bd0ee8ff0f68962b6a

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+* When the `clientID` message property is set, the key should have the same value as this property.
+* When the `clientID` message property is **not** set, the key should have the same value as the `topic` message property.
+
+{{< c8y-admon-info >}}
```
Probably not needed
```suggestion
```

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/382f350e65557ce4351ce3bd0ee8ff0f68962b6a

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+
+        // Publish a message to a single MQTT device
+        producer.newMessage()
+            .property("client", "demoClient")
```
```suggestion
            .property("clientID", "demoClient")
```

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+        // Publish a message to a single MQTT device
+        producer.newMessage()
+            .property("client", "demoClient")
+            .property("channel", "demoTopic")
```
```suggestion
            .property("topic", "demoTopic")
```

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+        
+        // Publish a message to all MQTT devices subscribed to a topic
+        producer.newMessage()
+            .property("client", "")
```
```suggestion
            .property("clientID", "")
```

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+        // Publish a message to all MQTT devices subscribed to a topic
+        producer.newMessage()
+            .property("client", "")
+            .property("channel", "demoTopic")
```
```suggestion
            .property("topic", "demoTopic")
```

schmiel-cumulocity commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+Any undelivered messages will be automatically deleted if they have been on the backlog for longer than the _time to live (TTL) limit_.
+This policy helps to limit overall resource usage and reduces the need to process outdated data after a prolonged disconnection of a consumer.
+
+No message will ever be deleted from the backlog unless it reaches its TTL limit.
```
Should we add something like this for more clarity?
```suggestion
No message will ever be deleted from the backlog unless it reaches its TTL limit or it has been processed by the MQTT Service.
```

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/382f350e65557ce4351ce3bd0ee8ff0f68962b6a

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,420 @@
+For a microservice client, you should use the credentials of the per-tenant [service user](/microservice-sdk/general-aspects/#users-and-roles) that will be passed to the microservice when the tenant is subscribed to it.
+For an external application user, you can use the credentials of any tenant user with the appropriate authorization roles assigned, as described below.
+The username must be in the form `<tenant>/<user>` where `<tenant>` is the tenant id, and `<user>` is a user within that tenant.
+
```
**_todo:_** As discussed offline, document the implications of enabling TFA for non-Microservice users, vis-à-vis the `devices` global role as the mitigating factor.
https://cumulocity.com/docs/authentication/basic-settings/#tfa-settings

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/8928289452867fad808de90b2ec0d82b9aa63be9

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Sławomir Chmiel <slawomir.chmiel@cumulocity.com>

scmi-c8y committed:
MTM-64360 Explain how to disable TFA for a user

scmi-c8y committed:
MTM-64360 Implement various review comments

scmi-c8y committed:
MTM-64360 Pre-requisities requested in review

scmi-c8y committed:
MTM-64360 Thoughts on handling Pulsar errors

rohitjoshi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
````diff
@@ -0,0 +1,465 @@
+```java
+        // Create a simple message listener that will log some details of
+        // each message received.
+        final MessageListener<String> listener = new MessageListener<String>() {
````
**_question:_** I believe this is yet to be fixed as per this [comment](https://github.com/Cumulocity-IoT/c8y-docs/pull/3642#discussion_r2379768121), right?  Same for example-code-3 (producer one) too.

scmi-c8y replied:
> **_question:_** I believe this is yet to be fixed as per this [comment](https://github.com/Cumulocity-IoT/c8y-docs/pull/3642#discussion_r2379768121), right? Same for example-code-3 (producer one) too.

Yes exactly. I need to run the sample against an environment with the tenant isolation changes to be completely sure that works, then I can merge it and update all the code snippets in this PR.

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/3eebfaade1bbd2994a48e40fbb42c3810b7f052d

rohitjoshi-c8y reviewed (approved):
Looks good :)

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+---
+
+{{< product-c8y-iot >}} microservices and external applications can consume messages published by devices connected to the MQTT Service, and publish messages back to those devices.
+To do this, your microservice or external application will connect to the {{< product-c8y-iot >}} Messaging Service, a modified deployment of [Apache Pulsar](https://pulsar.apache.org/), and use the Pulsar protocol to publish and consume MQTT messages.
```
It's true to say that it is indeed modified from upstream, but I don't know if that information is more helpful than potentially alarming
```suggestion
To do this, your microservice or external application will connect to the {{< product-c8y-iot >}} Messaging Service, a deployment of [Apache Pulsar](https://pulsar.apache.org/), and use the Pulsar protocol to publish and consume MQTT messages.
```

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+{{< /c8y-admon-info>}}
+
+The MQTT Service implements _device isolation_, meaning that MQTT devices connected to the MQTT Service **cannot** communicate directly with each other using the MQTT protocol.
+All inter-device communication will be managed by the client, as shown in the diagram.
```
Reduces the implication that there's any kind of magic automatic routing happening
```suggestion
All inter-device communication can/must be managed by the client, as shown in the diagram.
```

scmi-c8y replied:
How about "All inter-device communication must be managed explicitly by the client, as shown in the diagram"?

rthrippleton replied:
Yeah, that'd be great.

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/7111a538881cf9b7646e41f88365422b6c1227a0

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+2. The URL of the Messaging Service (Pulsar broker) in your {{< product-c8y-iot >}} environment.
+3. Credentials for a user in your tenant with permission to access MQTT Service data on the Messaging Service.
+
+Each of these pre-requisities is explained in detail below.
```
```suggestion
Each of these prerequisites is explained in detail below.
```

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+#### Example code
+
+The code snippet below shows how to use the Pulsar Java client library to connect to the Messaging Service with basic authentication.
+It assumes that the Pulsar URL is in the `C8Y_BASEURL_PULSAR` environment variable and that the tenant, username and password are provided on the command line.
```
This feels like quite a weird combination to me - some of your details as if you're a microservice, some as if you're an app. I get that we want to cover both, but doing it like this made me scratch my head (after taking off my "new user" hat, obviously).

I suppose we could have this sample work in both environments, e.g. `if C8Y_BASEURL_PULSAR then use C8Y_BASEURL_PULSAR and C8Y_USER and ...  else args` . But maybe that's too verbose. Honestly, I don't know the right answer!

scmi-c8y replied:
The example is explicitly not a microservice (to keep it as simple as possible) so really the only change I could make is to move the URL to the args. Is that worth doing at this stage? I'd rather not as it would mean changing the example and its documentation, but if you feel strongly about it I can do that.

rthrippleton replied:
I don't feel that strongly - it's weird, but not incorrect.

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Richard Thrippleton <richard.thrippleton@softwareag.com>

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+
+### Consuming messages from MQTT devices {#consuming-messages-from-mqtt-devices}
+
+All messages published by devices connected to the MQTT Service for a given tenant will be published to a _single_ Pulsar topic, identified by the URL `persistent://<tenant>/mqtt/from-device`.
```
As a dumb user who used to get confused between domain and id:
```suggestion
All messages published by devices connected to the MQTT Service for a given tenant will be published to a _single_ Pulsar topic, identified by the URL `persistent://<tenantID>/mqtt/from-device`.
```
(I think we also call it `tenantID` in other text in the MQTT service guide)

scmi-c8y replied:
Good catch, and there were a few more places where I could make the same change: https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/19b351c2eaf6471549d769a14c5269aa6994e404

scmi-c8y committed:
MTM-64360 tenant -> tenantID in a few places

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+
+To optimize resource usage, the Messaging Service imposes storage limits and a message time-to-live (TTL) on persistently stored messages.
+
+See the [service quotas](/service-terms/quotas/#mqtt-service) documentation for details on the default limits.
```
**question** - Will these actually be in place by the time we merge this doc? My understanding is that we've got an upcoming feature for this. Do we have pre-existing configurable quotas that are adequate?

scmi-c8y replied:
The ones discussed in this section (backlog limit and TTL) are already in place.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+   * Published messages may be rejected if other limits or quotas on the Messaging Service are reached.
+
+If your client is using the Java client library, almost all errors will be reported as a `PulsarClientException` thrown by a client library method.
+In some very rare cases a `SchemaSerializationException` runtime error might also be thrown, if the client has not used the `Schema.BYTES` schema and byte array payloads exclusively.
```
This was something I was unaware of - I assumed that Pulsar was always bytes on the wire, and that the schema was purely about Java-level conversions. Does that mean that our example code using `Schema.STRING` is potentially dangerous?

scmi-c8y replied:
Honestly, I'm not 100% certain how it works either. But after some discussion with Rohit & Slawek on the other PR we changed everything to use `<byte[]>` and `Schema.BYTES` to be safe.

rthrippleton commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+In general, it is not possible to recover from a fatal configuration or logic error in the client implementation.
+The client will need to be restarted after the error has been corrected.
+For transient errors, a strategy of retrying after a delay is usually appropriate.
+When an operation on a producer or a consumer has failed, we recommend deleting the failed producer or consumer object and creating a new one before retrying the operation.
```
There's a good chance that I'm a bit confused, but it sounds like a _lot_ of this is handled under the hood - see the first paragraph of https://pulsar.apache.org/docs/4.1.x/concepts-clients/ . So we don't want to be putting client authors to unnecessary effort.

scmi-c8y replied:
You may be right on the consumer side - someone with more experience than me will know better. But I do know that we do exactly what I'm suggesting in the core for producers as we've seen them get "stuck" and refuse to reconnect after earlier failures. It's probably overkill to always throw it away and create a new one, but it's a rare event and it should always work (eventually). I'd much rather the client authors put in some unnecessary effort than that they open a support ticket that we have to debug, TBH.

scmi-c8y replied:
@rohitjoshi-c8y Do you have any thoughts on this? I really don't know if we need to be as aggressive about replacing "dead" consumer objects as we have been with producers.

jato-c8y replied:
FWIW in Apama I never recreate a consumer or producer that may have failed to receive/produce. If either fail to create then I would retry until they do, but once initial connection is established I never delete them even on a fail to receive/produce. 
Certainly Apama tests that if a connection drops then on next attempt to receive the consumer is able to automatically re-establish the connection.
Also Apama is a c++ client which may differ slightly so your mileage may vary. 

I am not sure we care that we recommend the most efficient approach though, so belt and braces of destroy and create is probably a ok recommendation. The dev can use some brain/gpt power to make it more intelligent.

scmi-c8y replied:
I may well be describing a bug in the Java client library here! Although if it is, it's one that has persisted through many upgrades of the library. Anyway, I can try to make this a bit less scary... any suggestions on how to re-word it?

jato-c8y replied:
'When an operation on a producer or a consumer has failed, we recommend deleting the failed producer or consumer object and creating a new one before retrying the operation.'
\>\>\>
'When an operation on a producer or a consumer has failed, the error given may not be the easiest to identify a root cause, for a simple approach that would cover various scenarios, we would recommend deleting the failed producer or consumer object and creating a new one before retrying the operation, this should help to either identify the error at the earliest point or this may be enough to correct an underlying issue with those objects'

Maybe too wordy?

scmi-c8y replied:
Maybe a bit wordy, but it inspired me. Hopefully this is acceptable to everyone: https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/e9578b8ee365e081c325934e4ea5a90b7f5e74f1

scmi-c8y committed:
MTM-64360 apply a lightly modified review comment

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+
+#### Pulsar client library
+
+Open-source Pulsar client libraries are available for a number of different languages and protcols.
```
```suggestion
Open-source Pulsar client libraries are available for a number of different languages and protocols.
```

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+For an external application client, the URL has the general form `pulsar+ssl://<tenant_domain>:6651/`, where `<tenant_domain>` is the domain of your {{< product-c8y-iot >}} tenant, for example `my-tenant.cumulocity.com`.
+As implied by the `pulsar+ssl` protocol name, all external application client connections will use SSL/TLS security.
+Currently, only one-way TLS is supported; that is, the server will provide a certificate that can be verified by the client, but client certificates cannot be used.
+Implementing an external application client so that it reads the Pulsar URL from the `C8Y_BASEURL_PULSAR` environment variable will make it easier to develop client that can be deployed as either a microservice or an external application.
```
```suggestion
Implementing an external application client so that it reads the Pulsar URL from the `C8Y_BASEURL_PULSAR` environment variable will make it easier to develop a client that can be deployed as either a microservice or an external application.
```

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+The code snippet below shows how to use the Pulsar Java client library to connect to the Messaging Service with basic authentication.
+It assumes that the Pulsar URL is in the `C8Y_BASEURL_PULSAR` environment variable and that the tenant, username and password are provided on the command line.
+Note that the client library will not actually attempt to connect to the Pulsar server immediately when the `PulsarClient` object is created.
+In the interests of brevity and clarity, this example does no error handling.
```
```suggestion
In the interest of brevity and clarity, this example does no error handling.
```

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+#### Example code
+
+The code snippet below shows how to use the Pulsar Java client library to connect to the Messaging Service with basic authentication.
+It assumes that the Pulsar URL is in the `C8Y_BASEURL_PULSAR` environment variable and that the tenant, username and password are provided on the command line.
```
```suggestion
It assumes that the Pulsar URL is in the `C8Y_BASEURL_PULSAR` environment variable and that the tenant identifier, username and password are provided on the command line.
```

I think the comma is not required here.

scmi-c8y replied:
The comma was required - there are 3 separate command line args expected by the example. However it's much clearer if we say "tenant identifier" rather than just "tenant", so a good catch regardless.

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+| `tx.payloadFormatIndicator`<sup>(2)</sup> | NO                | Single byte with two permitted values, encoded as strings "0" and "1" | MQTT v5 Payload Format Indicator                     |
+| `tx.contentType`                          | NO                | String                                                                | MQTT v5 Content Type                                 |
+| `tx.responseTopic`                        | NO                | String                                                                | MQTT v5 Response Topic                               |
+| `tx.correlationData`                      | NO                | Sequence of bytes, encoded as a Base64 string                         | MQTT v5 Correlator Data                              |
```
```suggestion
| `tx.correlationData`                      | NO                | Sequence of bytes, encoded as a Base64 string                         | MQTT v5 Correlation Data                              |
```

Apama til you die?

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+| `tx.userProperties.<name>`                | NO                | String                                                                | MQTT v5 User Property with name `name`<sup>(3)</sup> |
+
+Notes:
+1. The `clientID` property can be omitted from a published message only in special case of a _broadcast_ message, described below in [broadcast messages](#broadcast-messages)
```
```suggestion
1. The `clientID` property can be omitted from a published message only in special case of a _broadcast_ message, described below in [broadcast messages](#broadcast-messages).
```

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+
+{{< c8y-admon-caution >}}
+Your client **must** be trusted to safely handle every message published by every device connected to the MQTT Service in your tenant.
+If untrusted users have access to your tenant, these users should **not** be permitted to upload microservices, nor to connect external application clients to the MQTT Service.
```
```suggestion
If untrusted users have access to your tenant, these users should **not** be permitted to upload microservices, nor to connect external application clients to the Messaging Service.
```

Right?

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+#### Example code
+
+The code snippet below shows how to use the Pulsar Java client library to consume messages from the MQTT Service `from-device` topic.
+It extends the previous example that set up the connection to the Pulsar server.
```
```suggestion
It extends the previous example that showed how to connect to the Messaging Service.
```

EH! This might be too picky you kind of say you don't connect to the Pulsar server in the above example code. Yes you do prepare the connection, but don't do the connection. (sorry)

jato-c8y replied:
Later you use 'set up the connection to the Pulsar server'
perhaps the 'tagline' of each example should be more prominent then easier to refer to in each subsequent example. 
Example code - Set up the connection to the Pulsar server
Example code - Create a message consumer
Example code - Produce a message

scmi-c8y replied:
Good idea, I will do that.

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/5bb2e153aeabbe7dafe2a303d7740c6cc017800d

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+| `topic`       | Name of the MQTT topic that the message should be published to       |
+
+If the `topic` property is empty or missing, the message will not be published to any MQTT client.
+The message will only be published to a client with an active subscription to the named MQTT topic.
```
```suggestion
The message will only be published to a device with an active subscription to the named MQTT topic.
```

I think client and device is interchanged in these paragraphs, might need a general once over to standardize.

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+        
+        // Publish a message to all MQTT devices subscribed to a topic
+        producer.newMessage()
+            .property("clientID", "")
```
This is an invalid message :/

'The `clientID` property is set but has an empty value.'

scmi-c8y replied:
Correct, and it will be fixed when I paste the final version of the sample in here (you need to look at the PR in cumulocity-examples if you care about the code)

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+
+#### Message time-to-live
+
+Any undelivered messages will be automatically deleted if they have been on the backlog for longer than the _time to live (TTL) limit_.
```
```suggestion
Any undelivered messages will be automatically deleted if they have been on the backlog for longer than the _time-to-live (TTL) limit_.
```

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+This recommendation also applies in the case of multiple customers, who do not mutually trust each other, sharing a single tenant.
+{{< /c8y-admon-caution >}}
+
+#### Durable subscriptions and acknowledgement
```
```suggestion
#### Durable subscriptions and message acknowledgement
```

scmi-c8y replied:
I didn't entirely agree with this one as it might it read like the acknowledgements themselves were durable. I hope my slight adjustment is OK?

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+1. Use the same subscription name every time the client connects a consumer.
+   A common _anti-pattern_ is to generate a random subscription name each time the client runs.
+   This will create a new subscription each time, but leave any previous subscriptions active with no consumers.
+   Eventually, the backlog quota limit for the topic will be reached, and no further messages will be deliver to the client.
```
```suggestion
   Eventually, the backlog quota limit for the topic will be reached, and no further messages will be delivered to the client.
```

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,465 @@
+
+### Example client
+
+A complete example client based on the code snippets above can be found in the [cumulocity-examples](https://github.com/Cumulocity-IoT/cumulocity-examples/mqtt-service/simple-pulsar-client) repository.
```
Just to confirm as I thought this may go live tomorrow, this example does not exist yet right? (link does not work)

scmi-c8y replied:
Correct again! The PR is approved and I just want to test it against an environment with the tenant isolation changes before I merge it.

jato-c8y reviewed (changes requested):
Really good document! 👏  Just some minor comments.

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Jack Tolfree <96070967+jato-c8y@users.noreply.github.com>

scmi-c8y committed:
MTM-64360 Titles for example code sections

scmi-c8y committed:
MTM-64360 Hedge our bets around this point...

scmi-c8y commented:
@BeateRixen @JanePorter I hope to have this PR ready to merge tomorrow - it just needs some final testing of the example code and dealing with any nit-picks from the reviewers. I'd like to have it go live on Thursday alongside the features that it describes. Can I get a code owner review before then please?

scmi-c8y committed:
MTM-64360 Update the example code snippets

schmiel-cumulocity commented on content/change-logs/platform-services/mqtt-service-0.9.21-pulsar-client-preview.md:
```diff
@@ -0,0 +1,40 @@
+This feature is in Public Preview and may be subject to change in the future.
+{{< /c8y-admon-preview >}}
+
+As [previously announced](/change-logs/#mqtt-service-0.9.6-device-isolation-api-change), when the {{< product-c8y-iot >}} [MQTT Service](/device-integration/mqtt-service/) transitions from Public Preview to General Availability (GA), the MQTT Service [Java client SDK](/device-integration/mqtt-service/#java-client) will no longer be supported.
```
As the [deprecation](https://cumulocity.com/docs/change-logs/#mqtt-service-0.9.x-sdk-java-announcement) announcement is already published shouldn't we link it here?

scmi-c8y replied:
https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/f266b2b3d0a2ddd380ac5c7b488bfa6f0f65ba6b

jato-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,496 @@
+                .send();
+            System.out.println("Sent message to all subscribed devices");
+
+            // Pause for a minute to allow some test messages to be consumed.
+            Thread.sleep(60 * 1000);
```
```suggestion
```

I don't think doc code should have sleeps.

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Jack Tolfree <96070967+jato-c8y@users.noreply.github.com>

jato-c8y committed:
Remove double spacing

scmi-c8y committed:
MTM-64360 Mention the SDK deprecation

scmi-c8y committed:
Merge branch 'feature/MTM-64360/mqtt-service-pulsar-client' of github.com:Cumulocity-IoT/c8y-docs into feature/MTM-64360/mqtt-service-pulsar-client

scmi-c8y committed:
MTM-64360 Fix URLs to example code

scmi-c8y committed:
MTM-64360 Table alignment nitpick

scmi-c8y committed:
MTM-64360 Apply suggestions from Copilot
* Align better with style guide
* Tighten prose (~10 lines shorter)
* Fix minor typos

scmi-c8y commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,493 @@
+Pulsar messages consist of a _payload_ and set of _properties_.
+
+The payload is a sequence of zero or more bytes, identical to the payload of the MQTT `PUBLISH` message that the Pulsar message corresponds to.
+It is the client's responsiblity to understand the format of the payloads produced and accepted by the MQTT devices it communicates with.
```
Fixed as part of https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/5e1863a8b8a530dd220b123a64f8a58dd587e898

scmi-c8y committed:
MTM-64360 Rename with version number

JanePorter reviewed (approved):
Very comprehensive and well written documentation, thanks guys.

BeateRixen commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,485 @@
+This documentation does not cover the publish-subscribe messaging concepts and architecture implemented by Pulsar, nor any features of the Pulsar client libraries beyond those needed to implement a simple MQTT Service client.
+To learn more about those subjects, refer to the [Pulsar product documentation](https://pulsar.apache.org/docs/4.0.x/).
+
+### Connecting to the Messaging Service
```
Please add the heading IDs throughout this file consistently.

scmi-c8y replied:
Done (hopefully I didn't miss any): https://github.com/Cumulocity-IoT/c8y-docs/pull/3642/commits/bba50039195eac243f108f0963bfee8ffe25e327

BeateRixen commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,485 @@
+### Example client
+
+A complete [example Java client](https://github.com/Cumulocity-IoT/cumulocity-examples/tree/develop/mqtt-service/java-simple-pulsar-client) based on the code snippets above can be found in the [cumulocity-examples](https://github.com/Cumulocity-IoT/cumulocity-examples/tree/develop/mqtt-service) repository.
+The `README.md` file provided with the example explains how to build and run it.
```
```suggestion
The *README.md* file provided with the example explains how to build and run it.
```

BeateRixen commented on content/device-integration/mqtt-service-bundle/pulsar-client.md:
```diff
@@ -0,0 +1,485 @@
+The `README.md` file provided with the example explains how to build and run it.
+
+The examples repository also contains a simple [Python MQTT client](https://github.com/Cumulocity-IoT/cumulocity-examples/tree/develop/mqtt-service/python-simple-mqtt-client) that can be used to simulate an MQTT device and test the operation of the Java client.
+See the `README.md` file included with the example for more details.
```
```suggestion
See the *README.md* file included with the example for more details.
```

BeateRixen commented:
> @BeateRixen Sorry for chasing but we need your approval as code owner to merge this PR. Thanks!

@scmi-c8y Looks great! Just two minor fixes and a comment on missing heading IDS which still must be added.

scmi-c8y committed:
Apply suggestions from code review

Co-authored-by: Beate Rixen <90445236+BeateRixen@users.noreply.github.com>

scmi-c8y committed:
MTM-64360  Add heading anchors everywhere
